### PR TITLE
[16.0][FIX]account_invoice_report_grouped_by_picking: do not show negative values with partially invoiced services

### PR DESCRIPTION
**Pull request by manuelregidor**
With this fix, partially invoiced services that come from sale order lines do not show negative values in invoices PDFs grouped by pickings.

Consider the following sale order:
grouped_1

and the following invoice, in which the amount of the invoiced service quantity is less than the quantity ordered:
grouped_2

Before this fix, there were two lines for the service product: in one of them, quantity is 5, which is the amount in the sale order; in the other one, quantity is -3, which is the amount that needs to be substracted from the other line in order to obtain the amount present in the invoice:
grouped_3

With the fix, there is only one line in the PDF that contains the amount that is actually invoiced.
grouped_4

### CODE REVIEW

**Comment by manuelregidor** [comment]
@HaraldPanten

**Review by HaraldPanten — APPROVED** [review]
Functional review. LGTM

**Review by etobella — APPROVED** [review]
Seems logic, can you add a test?

**Review by marcelsavegnago — APPROVED** [review]

**Review by CarlosRoca13 — APPROVED** [review]

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Sorry for stopping this, but I see that your change impacts both lines with product, consumable or service, so this may alter some of the previous behaviors. Should you limit it to services?

Also I would like to have a test exercising the specific problem for avoiding future regressions.

File: account_invoice_report_grouped_by_picking/models/account_move.py [hunk 1]
```diff
@@ -108,6 +108,10 @@ def lines_grouped_by_picking(self):
                         )
                         picking_dict.setdefault(key, 0)
                         qty = so_line.product_uom_qty
+                        if remaining_qty < so_line.product_uom_qty:
```
Comment (pedrobaeza) [inline account_invoice_report_grouped_by_picking/models/account_move.py]: and product type = service?

Comment (manuelregidor) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: I think the same logic should be followed for all products.

Take this sale order as an example:
group_1

This picking:
group_2

And this invoice:
group_3

This is the PDF:
group_4

The issue happens as well, as the amount in the invoice is less than the amount in the sale order plus the invoice line is not related to any stock.move.line as it has not been delivered.

Comment (pedrobaeza) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: @carlosdauden I think you implemented this system with negatives. Do you remember why?

Comment (HaraldPanten) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: @manuelregidor is right, we should consider services and consumables as well, as the same is happening in both situations.

Comment (pedrobaeza) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: Well, this is considering everything (including storable ones). That's why I'm asking my colleague if he remembers about this implementation, as I think he was the one doing it.

Comment (HaraldPanten) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: Yes, I said consumables as the example was with a consumable product, but the same is happening with any type of product. See this example:

Sale order with storable products:

Captura de Pantalla 2024-03-08 a las 9 50 34

Partial delivery order:

Captura de Pantalla 2024-03-08 a las 9 50 46

Invoicing the delivered products and just some of the other storables (invoicing by ordered quantities):

Captura de Pantalla 2024-03-08 a las 9 50 57

The report is strange like that:

Captura de Pantalla 2024-03-08 a las 9 51 10

**Comment by HaraldPanten** [comment]
Hi @pedrobaeza Any news about this? The error is in several versions, and it should be fixed.

**Comment by pedrobaeza** [comment]
@carlosdauden can you confirm that you see the change correct?

File: account_invoice_report_grouped_by_picking/models/account_move.py [hunk 2]
```diff
@@ -108,6 +108,10 @@ def lines_grouped_by_picking(self):
...
+                            qty = so_line.product_uom_qty - (
+                                so_line.product_uom_qty - remaining_qty
+                            )
```
Comment (pedrobaeza) [inline account_invoice_report_grouped_by_picking/models/account_move.py]: Matemáticamente, esto es lo mismo:

```suggestion
                            qty = remaining_qty
```

Comment (manuelregidor) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: done

File: account_invoice_report_grouped_by_picking/models/account_move.py [hunk 3]
```diff
@@ -108,6 +108,8 @@ def lines_grouped_by_picking(self):
                         )
                         picking_dict.setdefault(key, 0)
                         qty = so_line.product_uom_qty
+                        if remaining_qty < so_line.product_uom_qty:
+                            qty = remaining_qty
```
Comment (pedrobaeza) [inline account_invoice_report_grouped_by_picking/models/account_move.py]: And what about this more compact code?

```suggestion
                        qty = min(so_line.product_uom_qty, remaining_qty)
```

Comment (manuelregidor) [reply account_invoice_report_grouped_by_picking/models/account_move.py]: done

**Review by pedrobaeza — APPROVED** [review]
OK, let's go with this with the current consensus, and let's see if any problem is detected:

/ocabot merge patch

**Pull request merged**